### PR TITLE
WIP: Api ify bedwars + Add dragon bossbar 

### LINKED ISSUE #40: Database Improvements
**Is your feature request related to a problem? Please describe.**
Is your feature request related to a problem? Please describe.
BedWars1058 actually supports 2 types of databases (SQLite and MySQL).

**Describe the solution you'd like**
- Implement a database API with methods like initialization, stopping, saving stats, etc (could be an idea moving the Database.class from bedwars-plugin to bedwars-api) and implement new database types for saving stats/shops (especially NoSQL databases like MongoDB).
- Implement a method `BedWars#setDatabaseAdapter(IDatabase database)`

**Additional context**

### LINKED ISSUE #126: API Improvements
**Is your feature request related to a problem? Please describe.**
BedWars1058 enjoys a good amount of APIs that make the plugin remarkably customizable with addons, but in some parts of APIs it still lacks

**Describe the solution you'd like**
New APIs for greater customization:

- [x] Shop API (for example ShopManager or ShopIndex API)
- [x] Chat & Economy API (for economy also an event of moneyGain like PlayerXpGainEvent)
- [x] Database API, see #40
- [x] Stats API (Player Stats change event, stats interfaces and methods for creating new custom stats in addons with auto-implement in db)
- [x] Make MoneyGain & XpGain events cancellable
- [x] Upgrades API (like IUpgradesManager or a method to set an upgrade, like sharpened swords, to to a certain or maximum tier)
- [x] BedWars Addon system, see #166

### PR DESCRIPTION
**Pull request by tomkeuper**

### CODE REVIEW

**Review by Xxyuri2005xX — COMMENTED** [review]
we can proceed with the great merge

File: bedwars-api/src/main/java/com/tomkeuper/bedwars/api/addon/IAddonManager.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+     * Unload every addon
+     */
+    void unloadAddons();
+
+    /**
+     * Load every addon
+     */
+    void loadAddons();
+
+    void registerAddon(Addon addon);
```
Comment (SebastiaanDenHertog) [inline bedwars-api/src/main/java/com/tomkeuper/bedwars/api/addon/IAddonManager.java:69]: Comment

File: bedwars-api/src/main/java/com/tomkeuper/bedwars/api/arena/IArena.java [hunk 2]
```diff
@@ -485,4 +486,5 @@ public interface IArena {
     void setTeamAssigner(ITeamAssigner teamAssigner);
```
Comment (SebastiaanDenHertog) [inline bedwars-api/src/main/java/com/tomkeuper/bedwars/api/arena/IArena.java:691]: Comment

File: bedwars-api/src/main/java/com/tomkeuper/bedwars/api/server/RestoreAdapter.java [hunk 3]
```diff
@@ -101,4 +101,6 @@ public Plugin getOwner() {
      * Let them load on BedWars1058 main Thread so they will be converted before getting loaded.
      */
     public abstract void convertWorlds();
+
+    public abstract String getDisplayName();
```
Comment (SebastiaanDenHertog) [inline bedwars-api/src/main/java/com/tomkeuper/bedwars/api/server/RestoreAdapter.java:136]: Comment

**Review by SebastiaanDenHertog — APPROVED** [review]

**Comment by tomkeuper** [comment]
I'll quickly add the missing docs and then resolve conflicts. after that its ready for merge

**Review by porudev — APPROVED** [review]

**Review by KKiiya — APPROVED** [review]

**Pull request merged**